fcrepo-http-api/src/main/java/org/fcrepo/http/api/DefaultPathLockManager.java

```diff
@@ -0,0 +1,353 @@
+                    }
+                }
+
+            }
+            LOGGER.debug("Acquired all necessary path locks  (Thread {})", Thread.currentThread().getId());
+
+        }
+
+        private boolean tryAcquireAll() {
+            final List<ActivePath.PathScopedLock> acquired = new ArrayList<>();
+            for (final ActivePath.PathScopedLock lock : locks) {
+                if (lock.tryLock()) {
+                    acquired.add(lock);
+                } else {
+                    // roll back
+                    acquired.forEach(ActivePath.PathScopedLock::unlock);
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        /*
+         * This is the only method that removes paths from the pool
+         * of currently active paths that can be locked.
+         */
+        @Override
+        public void release() {
+            synchronized (DefaultPathLockManager.this) {
+                for (final ActivePath.PathScopedLock lock : locks) {
+                    lock.unlock();
+                    lock.getPath().threads.remove(Thread.currentThread());
+                    if (lock.getPath().threads.isEmpty()) {
+                        activePaths.remove(lock.getPath().path);
+                    }
+                }
+                if (deletePath != null) {
+                    LOGGER.trace("Thread {} releasing delete lock on path {}.",
+                            Thread.currentThread().getId(), deletePath);
+                    activeDeletePaths.remove(deletePath);
+                }
+                LOGGER.trace("Thread {} released locks.", Thread.currentThread().getId());
+                DefaultPathLockManager.this.notify();
+            }
+        }
+
+    }
+
+    /*
+     * This is the only method that adds paths to the pool of
+     * currently active paths that can be locked.
+     */
+    private synchronized ActivePath getActivePath(final String path) {
+        ActivePath activePath = activePaths.get(path);
+        if (activePath == null) {
+            activePath = new ActivePath(path);
+            activePaths.put(path, activePath);
+        }
+        return activePath;
+    }
+
+    private boolean isOrIsDescendantOf(final String possibleDescendant, final String path) {
+        return path.equals(possibleDescendant) || possibleDescendant.startsWith(path + "/");
+    }
+
+    private String getParentPath(final String path) {
+        if (path.indexOf('/') == -1) {
+            return null;
+        }
+        return path.substring(0, path.lastIndexOf('/'));
+    }
+
+    @VisibleForTesting
+    static String normalizePath(final String path) {
+        if (path.endsWith("/")) {
+            return path.substring(0, path.length() - 1);
+        } else {
+            return path;
+        }
+    }
+
+    @Override
+    public AcquiredLock lockForRead(final String path, final Session session, final NodeService nodeService)
```

Comment:
It's been a while, but I think my reason was primarily because I could imagine an alternative implementation that might need it.